packages/esm-patient-allergies-app/src/allergies/allergies-form/allergy-form.component.tsx

```diff
@@ -0,0 +1,298 @@
+  const [comment, setComment] = React.useState<string>();
+  const [patientReactions, setPatientReactions] = React.useState<Array<string>>([]);
+  const [selectedAllergen, setSelectedAllergen] = React.useState<string>();
+  const [severityOfReaction, setSeverityOfReaction] = React.useState<string>();
+  const [dateOfOnset, setDateOfOnset] = React.useState<string | Date>();
+  const [status, setStatus] = React.useState<StateTypes>(StateTypes.PENDING);
+  const [allergenType, setAllergenType] = React.useState<AllergenType>(AllergenType.DRUG);
+  const [otherReaction, setOtherReaction] = React.useState<string>();
+  const [error, setError] = React.useState();
+
+  React.useEffect(() => {
+    if (drugAllergenUuid && foodAllergenUuid && environmentalAllergenUuid) {
+      const sub = fetchAllergensAndReaction([
+        drugAllergenUuid,
+        foodAllergenUuid,
+        environmentalAllergenUuid,
+        allergyReactionUuid,
+      ]).subscribe(
+        (data) => {
+          setAllergenAndReactions(data);
+          setStatus(StateTypes.RESOLVED);
+        },
+        (error) => {
+          setError(error);
+          createErrorHandler();
+          setStatus(StateTypes.ERROR);
+        },
+      );
+      return () => sub.unsubscribe();
+    }
+  }, [drugAllergenUuid, environmentalAllergenUuid, foodAllergenUuid, allergyReactionUuid]);
+
+  const handlePatientReactionChange = React.useCallback(
+    (value: boolean, id: string, event: React.ChangeEvent<HTMLInputElement>) => {
+      value
+        ? setPatientReactions((prevState) => [...prevState, id])
+        : setPatientReactions((prevState) => prevState.filter((reaction) => reaction !== id));
+    },
+    [],
+  );
+
+  const handleAllergenTypeChange = React.useCallback((index: number) => {
+    if (index === 0) setAllergenType(AllergenType.DRUG);
+    if (index === 1) setAllergenType(AllergenType.FOOD);
+    if (index === 2) setAllergenType(AllergenType.ENVIRONMENT);
+  }, []);
+
+  const handleSavePatientAllergy = React.useCallback(() => {
+    const restApiPayLoad = {
+      allergen: {
+        allergenType: allergenType,
+        codedAllergen: {
+          uuid: selectedAllergen,
+        },
+      },
+      severity: {
+        uuid: severityOfReaction,
+      },
+      comment: comment,
+      reactions: patientReactions?.map((reaction) => {
+        return reaction === otherReactionUuid
+          ? { reaction: { uuid: reaction }, reactionNonCoded: otherReaction }
+          : { reaction: { uuid: reaction } };
+      }),
+    };
+
+    const ac = new AbortController();
+    savePatientAllergy(restApiPayLoad, patientUuid, ac).then((response) => {
+      if (response.status === 201) {
+        closeWorkspace();
+
+        showToast({
+          kind: 'success',
+          description: t('allergySaved', 'Allergy saved successfully'),
+        });
+      }
+      (err) => {
+        createErrorHandler();
+        setStatus(StateTypes.ERROR);
```

Comment:
That we have multiple setX calls here is a sign of a join state. Sure you want to have it like that? If so, a `useReducer` would make more sense.